### PR TITLE
Added a shareable wrapper for NAL stack implementations

### LINKED ISSUE #43: Shared NAL
Once #42 is merged, the NAL traits in this crate will no-longer be sharable on their own.  It is necessary to create a new layer that can handle the sharing internally.

This can be split into two different types of sharing: single-thread (where RefCell can be used), and multi-threaded.  Both can draw inspiration from [Rahix/shared-bus](/Rahix/shared-bus).

I've drafted a simple RefCell-based wrapper here, inspired by the shared-bus Simple Manager.  It's suitable for sharing access to the network stack in a single-threaded context.  However @ryan-summers you mentioned that you think there may be issues with this approach?  Would you mind clarifying?

More work will be necessary to create a sharing layer that's either Sync or Send.

```rust
use core::cell::RefCell;
use embedded_nal::{nb, SocketAddr, TcpClientStack, TcpFullStack, UdpClientStack, UdpFullStack};

pub struct SharedNalManager<T> {
    driver: RefCell<T>,
}

impl<T> SharedNalManager<T> {
    pub fn new(driver: T) -> Self {
        SharedNalManager {
            driver: RefCell::new(driver),
        }
    }

    pub fn acquire(&self) -> SharedNal<T> {
        SharedNal {
            driver: &self.driver,
        }
    }
}

pub struct SharedNal<'a, T> {
    driver: &'a RefCell<T>,
}

macro_rules! forward {
    ($func:ident($($v:ident: $IT:ty),*) -> $T:ty) => {
        fn $func(&mut self, $($v: $IT),*) -> $T {
            self.driver.borrow_mut().$func($($v),*)
        }
    }
}

impl<'a, T: UdpClientStack> UdpClientStack for SharedNal<'a, T> {
    type Error = T::Error;
    type UdpSocket = T::UdpSocket;

    forward! {socket() -> Result<Self::UdpSocket, Self::Error>}
    forward! {connect(socket: &mut Self::UdpSocket, address: SocketAddr) -> Result<(), Self::Error>}
    forward! {send(socket: &mut Self::UdpSocket, data: &[u8]) -> Result<(), nb::Error<<T as embedded_nal::UdpClientStack>::Error>>}
    forward! {receive(socket: &mut Self::UdpSocket, data: &mut [u8]) -> Result<(usize, SocketAddr), nb::Error<<T as UdpClientStack>::Error>>}
    forward! {close(socket: Self::UdpSocket) -> Result<(), Self::Error>}
}

impl<'a, T: UdpFullStack> UdpFullStack for SharedNal<'a, T> {
    forward! {bind(socket: &mut Self::UdpSocket, local_port: u16) -> Result<(), Self::Error>}
    forward! {send_to(socket: &mut Self::UdpSocket, remote: SocketAddr, buffer: &[u8]) -> Result<(), nb::Error<<T as UdpClientStack>::Error>>}
}

impl<'a, T: TcpClientStack> TcpClientStack for SharedNal<'a, T> {
    type TcpSocket = T::TcpSocket;
    type Error = T::Error;

    forward! {socket() -> Result<Self::TcpSocket, Self::Error>}
    forward! {connect(socket: &mut Self::TcpSocket, address: SocketAddr) -> Result<(), nb::Error<<T as TcpClientStack>::Error>>}
    forward! {send(socket: &mut Self::TcpSocket, data: &[u8]) -> Result<usize, nb::Error<<T as embedded_nal::TcpClientStack>::Error>>}
    forward! {receive(socket: &mut Self::TcpSocket, data: &mut [u8]) -> Result<usize, nb::Error<<T as TcpClientStack>::Error>>}
    forward! {is_connected(socket: &Self::TcpSocket) -> Result<bool, Self::Error>}
    forward! {close(socket: Self::TcpSocket) -> Result<(), Self::Error>}
}

impl<'a, T: TcpFullStack> TcpFullStack for SharedNal<'a, T> {
    forward! {bind(socket: &mut Self::TcpSocket, port: u16) -> Result<(), <T as TcpClientStack>::Error>}
    forward! {listen(socket: &mut Self::TcpSocket) -> Result<(), <T as TcpClientStack>::Error>}
    forward! {accept(socket: &mut Self::TcpSocket) -> Result<(<T as TcpClientStack>::TcpSocket, SocketAddr), nb::Error<<T as TcpClientStack>::Error>>}
}
```

### PR DESCRIPTION
**Pull request by jonahbron**
Ended up with some slightly different names that made more sense to me in the context of the traits and docs.  Open to feedback.

/fixes #43

### CODE REVIEW

File: src/stack/share.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+macro_rules! forward {
+    ($func:ident($($v:ident: $IT:ty),*) -> $T:ty) => {
+        fn $func(&mut self, $($v: $IT),*) -> $T {
+            self.stack.borrow_mut().$func($($v),*)
+        }
+    }
+}
+
+impl<'a, T: UdpClientStack> UdpClientStack for SharedStack<'a, T> {
```
Comment (ryan-summers) [inline src/stack/share.rs]: I would recommend opting for the `where` clause instead of `T: UdpClientStack` - I find it more readable and easy to follow.

Comment (ryan-summers) [inline src/stack/share.rs]: This feels like something we may need to have a test for to verify that we can make multiple copies of it.

File: src/stack/share.rs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	/// Returns a shared reference to the driver that can be used as a first-class implementation.
+	pub fn acquire(&self) -> SharedStack<T> {
+		SharedStack { stack: &self.stack }
+	}
+}
+
+/// Single-thread shared reference to an internal network stack implementation.
+///
+/// This can only be created by calling [`SharableStack::acquire()`]
+pub struct SharedStack<'a, T> {
```
Comment (ryan-summers) [inline src/stack/share.rs:86]: I think we could provide some sort of example usage here?

File: src/stack/share.rs (hunk 1 above)
Comment (jonahbron) [reply src/stack/share.rs]: With the magic of doc tests, I've provided both a usage example _and_ a sharing test at once :smile:

**Comment by jonahbron** [comment]
The rust doc.  This works with doc tests too.

image

**Comment by jonahbron** [comment]
@Rahix Since we've been referencing your work on this, seems logical that you be able to offer feedback.

**Comment by jonahbron** [comment]
Ping again for feedback on this.

**Review by ryan-summers — APPROVED** [review]
LGTM. Thanks for the ping and the work on this! I'm going to be prepping a new release soon as well.

**Comment by ryan-summers** [comment]
@jonahbron Please add a CHANGELOG entry and I'll merge this

**Comment by Rahix** [comment]
Hey!

Sorry, I didn't get around to taking a look yet and quite frankly the thread got buried in my inbox so I forgot :/

Basically there isn't much to get wrong though as long as no unsafe or multi-tasking is involved (multi-tasking is prevented by `!Send` on the `RefCell`).  So I think the implementation is solid :)

One idea I'd like to put out there (though this could always be considered later on as well):  Maybe it would make sense to add this support into `shared-bus` instead?  This way, you could benefit from all the other mutex implementations over there as well, with zero additional effort.

**Review by MathiasKoch — APPROVED** [review]
Only thing is if it would be worth having the option to somehow catch `borrow_mut()` before panic ? `try_borrow_mut()`? But i think it would screw up the error types quite alot.

**Comment by Rahix** [comment]
@MathiasKoch, I do not think there is much to be gained.  The only reason why `borrow_mut()` could panic would be if the stack somehow accesses another `SharedStack` reference to itself.  I would argue that this is a serious programming error and not a failure condition that should be handled in production.

**Comment by MathiasKoch** [comment]
Then i think all we are missing is a changelog entry from @jonahbron before we can merge this and do a new 0.3 release.

> One idea I'd like to put out there (though this could always be considered later on as well): Maybe it would make sense to add this support into shared-bus instead? This way, you could benefit from all the other mutex implementations over there as well, with zero additional effort.

I think this would be nice eventually, but might be much easier once the soundness of this design is proven?

**Comment by jonahbron** [comment]
@ryan-summers changelog added

**Review by ryan-summers — APPROVED** [review]

**Pull request merged**